nabs/callbacks.py

```diff
@@ -0,0 +1,166 @@
+
+import pandas as pd
+from bluesky.callbacks.core import CallbackBase, make_class_safe
+
+logger = logging.getLogger(__name__)
+
+
+@make_class_safe(logger=logger)
+class ELogPoster(CallbackBase):
+    """
+    Callback to post messages to the elog, primarily related to plan setup and
+    results. This callback relies on and must be instantiated after the
+    BestEffortCallback and HutchELog.
+
+    To subscribe:
+
+    .. code-block:: python
+
+        elogc = ELogPoster(bec, elog)
+        elogc_uid = RE.subscribe(elogc)
+
+    To enable posting for a specific run:
+
+    .. code-block:: python
+
+        RE(plan(), post=True)
+
+    Parameters
+    ----------
+    bec : BestEffortCallback
+        Instance of BestEffortCallback subscribed to the current RunEngine
+
+    elog : HutchELog
+        Instance of HutchELog connected to the desired logbook
+    """
+    _FMTLOOKUP = {'string': '{:.5s}',
+                  'float': '{:.2f}',
+                  'number': '{:.3g}',
+                  'integer': '{:d}'}
+
+    _html_head = """<!DOCTYPE html>
+<html>
+<head>
+<style>
+table {
+  font-family: arial, sans-serif;
+  border-collapse: collapse;
+  width: 100%;
+}
+
+td, th {
+  border: 1px solid #dddddd;
+  text-align: left;
+  padding: 8px;
+}
+
+tr:nth-child(even) {
+  background-color: #dddddd;
+}
+</style>
+</head>
+<body>
+"""
+
+    _html_tail = "</body> </html>"
+
+    def __init__(self, elog, ipython):
+        self._elog = elog
+        self._ipython = ipython
+        self._data = dict()
+        self._descriptors = set()
+        self._data_keys = []
+        self._format_info = dict()
+
+        # TO-DO:
+        # - add granular switches as optional arguments
+        # - add error handling for misspelled keys
+        # - add default lists for start doc keys, elog tags?
+        # - handling if data key collides with seq_num, time?
+        # - consider pulling precision from data
+
+    def start(self, doc):
+        """ Post plan information on start document"""
+        self._send_post = doc.get('post', self._elog.enable_run_posts)
+
+        # Clean up before starting new run
+        self._data = dict()
+        self._descriptors = set()
+        self._data_keys = []
+        self._format_info = dict()
+
+        if self._send_post:
+            # Grab last ipython input
+            run_info = self._ipython.user_ns["In"][-1]
+            logger.info("Posting run start information to elog")
+            self._elog.post(run_info, tags=['plan_info', 'RE'])
+
+        super().start(doc)
+
+    def descriptor(self, doc):
+        """ Initialize table information """
+        # these always exist
+        self._data = dict(seq_num=[], time=[])
+        for k in doc['data_keys']:
+            dk_entry = doc['data_keys'][k]
+
+            if dk_entry['dtype'] not in self._FMTLOOKUP:
+                logger.warn(f'skipping {k}, format not recognized')
+                continue
+
+            self._data.update({k: []})
```

Comment:
`update` is a bit awkward for a single key, would suggest:
```suggestion
            self._data[k] = []
```